src/compiler/checker.ts

```diff
@@ -6334,11 +6348,22 @@ module ts {
             return getSignatureInstantiation(signature, getInferredTypes(context));
         }
 
-        function inferTypeArguments(signature: Signature, args: Expression[], excludeArgument: boolean[]): InferenceContext {
+        function inferTypeArguments(signature: Signature, args: Expression[], excludeArgument: boolean[], context: InferenceContext): void {
             let typeParameters = signature.typeParameters;
-            let context = createInferenceContext(typeParameters, /*inferUnionTypes*/ false);
             let inferenceMapper = createInferenceMapper(context);
 
+            // Clear out all the inference results from the last time inferTypeArguments was called on this context
+            for (let i = 0; i < typeParameters.length; i++) {
+                // As an optimization, we don't have to clear (and later recompute) inferred types

```

Comment:
It's not an optimization that we reset it. It's an optimization that we skip resetting it if the type parameter was already fixed. It would be just as correct to reset all of them. But then we'd be repeating the same work for the type parameters that were fixed. Namely the work done by getInferredType.